### ISSUE TITLE
RFC(slice-machine-ui): use/configure custom mocks

### ISSUE THREAD

**Issue by hypervillain** [prismicio/slice-machine#58]
As most users have noted, created mocks are pretty limited at the moment, if not existent. This is a pretty tough subject for us, as we both want to:

* help developers test components quickly, without having to "care" about actual content
* create screenshots relevant enough to be understood by their content writers

I believe the chosen solution should take care of these 2 problems, that belong to different moments of the same flow. I've been building a small prototype that tries to address them:

Basically, it expects a `.slicemachine/mocks.json` file (that I create manually atm but could be handled by the builder) that can alter the way a mock will be generated. This configuration occurs per field and depends on the widget type the field points to.

A minimum example for a library containing a `CallToAction` slice with 2 fields would look like his:

```JSON
{
  "CallToAction": {
    "title": {
      "config": {
		"sentencesPerParagraph": {
          "min": 12,
          "max": 44
        },
        "patternType": "PARAGRAPH_WITH_HEADING"
      }
    },
    "bool": {
      "content": true
    }
  }
}
````

For each widget type, 2 keys are available: `config` and `content`, where `content` takes priority on `config`. Of course, depending on widget type, different configurations are available.

So, taking this case as example: https://community.prismic.io/t/change-mock-images/2478/5, you could end up with one of these:

```JSON
"image": {
      "content": {
		"alt": "My own alternative text",
        "src": "https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885__340.jpg"
      }
    }

````
👆 One image defined, thumbnails generated automatically

```JSON
"image": {
      "config": { "imageType": "ICON" }
}

````
👆 Image of type icon, picked randomly

----
For most field types, I don't see a real advantage in defining a `config` as these fields are pretty straightforward. For example, providing a configuration object to Boolean fields seems a bit overkill.

For others, I've started documenting things here: https://www.notion.so/prismic/Mocks-by-Widget-a8c441c7054d4c0598c265fc99302f0b

**Comment by lihbr** [comment]
Great job! I like the config interface, easily editable programmatically by the UI afterward.

It'll be nice to separate mock configs on a slice/library level, becomes easier to understand and troubleshoot? Also it'll be great to take into account that you can have different types of mocks for a same variation of a given slice (one with everything filled, one with some field missing, one with this thing activated, one with a really long text, etc.) could become really handy to programmatically generate stories out of them afterward :)

Basically here's the interface I have in mind:
```jsonc
/* .slicemachine/libraries/<lib_name>/<slice_name>/mocks.json */
{
  "<variation_name_a>": [
    {
      "name": "<name_of this mock>" /* For example: "Filled", "Title Missing", etc. */
      "primary": { ...primaryFieldsMockConfig },
      "items": { ...itemsFieldsMockConfig }
    },
    { ...anotherMockConfigForVariationA }
  ],
  "<variation_name_b>": [ ...anotherArrayOfMockConfigsForVariationB ],
  ...otherVariations
}
```

**Comment by Duaner** [comment]
For images, Will it be okay if you could pass a specific unsplash URL? We could take care of dynamically resizing, cropping the image thanks to their API. 

If it's a custom image, we should probably host it in order to be able to crop it for the mocks and fit the constraints set in the builder

**Comment by samuelhorn** [comment]
I would love to pass an unsplash id, and slicemachine took care of the rest with unsplash api like:
`https://source.unsplash.com/WLUHO9A_xik/1600x900`.

Thats what i do for now in my customMock, but slicemachine doesn't seem to like it sometimes, as in some slices it cant take a screenshot with an unsplash image, and hence can not push to prismic either, but in some slices it works. Very strange.

**Comment by lihbr** [comment]
Yes, it'll be nice to be able to input any image URL and pipe it to Prismic imgix bucket @Duaner

**Comment by Duaner** [comment]
@lihbr any images could be abused a bit, but either specify an image in a prismic repo or an image coming from unsplash that uses imgix ? Wdyt ?

**Comment by lihbr** [comment]
Of course if that's much simpler security wise let's first scope it to images from trusted sources ☺️
I thought imgix had a feature like Cloudinary fetch thing that does all the security checks on their end: https://cloudinary.com/documentation/fetch_remote_images#remote_image_fetch_url

**Comment by davidparys** [comment]
That would be great. Currently, I am using storybook with Chromatic as a design system, each time I edit the model, the mocks are also changed and I have to say it's not ideal.. 

![Screenshot 2020-12-23 at 10 25 14](https://user-images.githubusercontent.com/56082497/102981380-27d19180-4509-11eb-8733-30b9ded3d153.png)

**Comment by ReeceM** [comment]
Not sure if this is the right one to post this, I tried the custom mocks stuff just now, I noticed that the tabs are not highlighted when clicking on it, it's not obvious that you have clicked on the right one as they stay the same state.

Another question, would it be more practical to have as the mock options only the valid choices? It might require less cognitive power and action to pick between the only two options, rather than two possible and two blank ones?

For example, I have attached a comparative image below of what I am talking about, I know it's probably just UI, but maybe the fact that there might be a few options on the mocks, it might just be cluttered (not sure of the extent planed)

### Simplified view
<img width="934" alt="Screen Shot 2021-01-29 at 20 23 18" src="https://user-images.githubusercontent.com/2767904/106312926-17cb7100-6270-11eb-9b98-8908eb62f35e.png">

### The original
<img width="933" alt="Screen Shot 2021-01-29 at 20 22 59" src="https://user-images.githubusercontent.com/2767904/106312937-1ac66180-6270-11eb-998b-3f854e2a1f94.png">

Also, I am not sure. But using the mock for images opens up a popup window, on safari if the pop ups are blocked and you miss it, clicking the button again won't load up again. Not sure what to look for on debugging that bit right now. 🤷‍♂️

**Comment by ReeceM** [comment]
Hi Hugo and Lucie, I think I might have found an edge case issue that has interaction with the Variation and Mocks together. 

I just am in between things trying to trace down the exact lines.

So don't want to just open a blank issue, or is it okay if I do that and then update it with the lines and console output?

**Comment by ReeceM** [comment]
### Additional Feedback

The mocks have been working great that I have been working with them again. 

I have an interesting thought. So for paragraphs there is the option to pick the number of blocks like when it is for rich text.

But I have notice, and there is the pattern where the `Text Content` type is used for link button text and navbar links. No considering the normal design pattern is 1 - 2 words on a navbar, there are un-needed layout issues when a `Text Content` mock has 3 - 4 words, it causes some issues if you try use repeated types. 

It would be nice to be able to say, give me random words, but no more than 2. (Or min max on it, possible?)

But then default to normal behaviour when no additional changes are made

What I have noticed would be nice, the **`Repeatable zone`** doesn't have a way to adjust the number of repeated instances.

I think in some cases it might be nice to cap the number of repeated items that are mocked, as when it keeps changing, while doing the design layout, it goes from 6 to 2 to 3 back to 6 and so on, so there isn't consistency, at least there.

I might be then making the point of repeatable zones pointless at that point 🤣

**Comment by lihbr** [comment]
Thanks Reece for the additional feedback! Indeed that would be great and will most likely be part of our 2nd iteration on mocks!

**Comment by hypervillain** [comment]
For RichText, you can manually add a "content" key to your field in `.slicemachine/mocks.json`:

```javascript
{
	"CallToAction: {
			"myRichText": {
				"content": {
					"text": "Two Words",
					"type": "heading2"
				}
			}
	}
}
````
Internally, we have everything available to define min-max number of paragraphs and words so we'll discuss the possibility to make them available through the interface really soon

**Comment by ReeceM** [comment]
> Hi Hugo and Lucie, I think I might have found an edge case issue that has interaction with the Variation and Mocks together.
> 
> I just am in between things trying to trace down the exact lines.
> 
> So don't want to just open a blank issue, or is it okay if I do that and then update it with the lines and console output?

Not sure if this was answered, but should I open a new issue for what I have found ? It might be a process flow issue or a bug depending on how you look at it.

**Comment by lihbr** [comment]
> Not sure if this was answered, but should I open a new issue for what I have found ? It might be a process flow issue or a bug depending on how you look at it.

Sure! Go ahead ☺️ Thanks again for all your contributions ♥

**Comment by ReeceM** [comment]
Hey @lihbr it must have been fixed then between the `alpha.6` version and the release of 0.0.44, awesome then \:)

Tried the update when checking if it was a version issue with the alpha (given I guess)

It was this error that doesn't come up now: `[sliceName].js:1 Warning: An unhandled error was caught from submitForm() TypeError: Cannot convert undefined or null to object`
